[quant][graphmode][refactor] Separate preprocess step for insertObserver

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #33650 [not4land] debugging
* #33533 [not4land] Temporary workaround to get quantized ops
* #33481 [quant][graphmode] Handling ops doesn't require observation in insertObservers
* #33532 [quant][graphmode] Add quantization pattern for quantized::add_relu
* #33279 [quant][graphmode] Add quantized conv2d-relu fusion pattern
* #32816 [quant][graphmode] Add add_relu pattern in skip values
* #32814 [quant][graphmode] Skip quantizing input and output in matched module
* #33277 [WIP][quant][graphmode] Observing input/output values in callsite
* #33649 [quant][graphmode][refactor] Factor out getInvokedMethod
* #33276 [quant][graphmode][refactor] Move check for weight outside of insertObserverFor
* #33275 [quant][graphmode][refactor] Move values_to_skip check inside valueNeedsToBeQuantized
* #33274 [quant][graphmode][refactor] Simplify signature for insertObserverFor
* #33273 [quant][graphmode][refactor] Checks for bias and weight
* #33173 [quant][graphmode] Add a pass to handle ops that doesn't require observation
* #33531 [quant][graphmode] Replicate dequantize nodes
* #33172 [quant][graphmode][refactor] Factor out insertDequantCall
* #33171 [quant][graphmode] refactor nodeQuantizable
* **#32813 [quant][graphmode][refactor] Separate preprocess step for insertObserver**
* #32812 [quant][graphmode][refactor] Change signature of getModuleAccessPath
* #32809 [quant][graphmode][refactor] Move the check for qconfig inside insertObserver call
* #32379 [quant][graphmode] FoldConvBatchNorm2d support shared ClassTypes

Summary:
We need to separate the step to make the logic more clear
and also to find all the values we want to skip in advance
without the interference of inserted observers

Test Plan:
.

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D20087841](https://our.internmc.facebook.com/intern/diff/D20087841)

jerryzh168 committed:
[quant][graphmode][refactor] Separate preprocess step for insertObserver

Summary:
We need to separate the step to make the logic more clear
and also to find all the values we want to skip in advance
without the interference of inserted observers

Test Plan:
.

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit 7ee522a9:


**None of the build failures appear to be your fault**.

* **1/1** [broken upstream](https://dr.pytorch.org/master-timeline.html?count=1&sha1=cccf5e7011f689c8dea499a1ab881900152e2575&should_suppress_scheduled_builds=true&should_suppress_fully_successful_columns=true&max_columns_suppress_successful=35) at merge base cccf5e70 since Jan 29
  <details>
  <summary>
  Please rebase on the <code>viable/strict</code> branch (expand for instructions)
  </summary>
  
  If your commit is newer than `viable/strict`, you can try basing on an older, stable commit:
  ```
  git fetch origin viable/strict
  git rebase --onto viable/strict $(git merge-base origin/master HEAD)
  ```
  If your commit is older than `viable/strict`:
  ```
  git fetch origin viable/strict
  git rebase viable/strict
  ```
  Check out the [recency history](https://dr.pytorch.org/master-viable-commits.html) of this "viable master" tracking branch.
  
  </details>
  


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=7ee522a967c01aa9b3bdc4cebb032865cdec830d).

### :construction: 1 upstream failure recognized by patterns:
These builds matched patterns, but were probably **caused by upstream breakages**:
* [pytorch_linux_backward_compatibility_check_test](https://dr.pytorch.org/build-details.html?build_id=93011380) from Jan 29



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

jerryzh168 committed:
Update on "[quant][graphmode][refactor] Separate preprocess step for insertObserver"


Summary:
We need to separate the step to make the logic more clear
and also to find all the values we want to skip in advance
without the interference of inserted observers

Test Plan:
.

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][refactor] Separate preprocess step for insertObserver"


Summary:
We need to separate the step to make the logic more clear
and also to find all the values we want to skip in advance
without the interference of inserted observers

Test Plan:
.

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][refactor] Separate preprocess step for insertObserver"


Summary:
We need to separate the step to make the logic more clear
and also to find all the values we want to skip in advance
without the interference of inserted observers

Test Plan:
.

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][refactor] Separate preprocess step for insertObserver"


Summary:
We need to separate the step to make the logic more clear
and also to find all the values we want to skip in advance
without the interference of inserted observers

Test Plan:
.

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][refactor] Separate preprocess step for insertObserver"


Summary:
We need to separate the step to make the logic more clear
and also to find all the values we want to skip in advance
without the interference of inserted observers

Test Plan:
.

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][refactor] Separate preprocess step for insertObserver"


Summary:
We need to separate the step to make the logic more clear
and also to find all the values we want to skip in advance
without the interference of inserted observers

Test Plan:
.

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][refactor] Separate preprocess step for insertObserver"


Summary:
We need to separate the step to make the logic more clear
and also to find all the values we want to skip in advance
without the interference of inserted observers

Test Plan:
.

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][refactor] Separate preprocess step for insertObserver"


Summary:
We need to separate the step to make the logic more clear
and also to find all the values we want to skip in advance
without the interference of inserted observers

Test Plan:
.

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][refactor] Separate preprocess step for insertObserver"


Summary:
We need to separate the step to make the logic more clear
and also to find all the values we want to skip in advance
without the interference of inserted observers

Test Plan:
.

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][refactor] Separate preprocess step for insertObserver"


Summary:
We need to separate the step to make the logic more clear
and also to find all the values we want to skip in advance
without the interference of inserted observers

Test Plan:
.

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][refactor] Separate preprocess step for insertObserver"


Summary:
We need to separate the step to make the logic more clear
and also to find all the values we want to skip in advance
without the interference of inserted observers

Test Plan:
.

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

raghuramank100 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -671,6 +676,28 @@ void InsertObserversHelper::addIntermediateValuesToSkipObserver(
+  ConstantPooling(graph);
+  ConstantPropagation(graph);
+  // must do constant propagation first before replacement
+  replaceConvolutionWithConv2d(graph);
```
Do these steps need to happen prior to Conv+ BN fusion? It feels like we should have the passes in the order:
 - PreProcess -> FoldConvBN->DeDupModules->InsertObservers.

Wher

jerryzh168 committed:
Update on "[quant][graphmode][refactor] Separate preprocess step for insertObserver"


Summary:
We need to separate the step to make the logic more clear
and also to find all the values we want to skip in advance
without the interference of inserted observers

Test Plan:
.

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][refactor] Separate preprocess step for insertObserver"


Summary:
We need to separate the step to make the logic more clear
and also to find all the values we want to skip in advance
without the interference of inserted observers

Test Plan:
.

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][refactor] Separate preprocess step for insertObserver"


Summary:
We need to separate the step to make the logic more clear
and also to find all the values we want to skip in advance
without the interference of inserted observers

Test Plan:
.

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

dreiss reviewed (approved):
Would like to see a little more in the test plan, even if it's just the name of the existing test that covers this.  That way, anyone looking at this code in the future will have a better idea of how to test changes to it.

jerryzh168 committed:
Update on "[quant][graphmode][refactor] Separate preprocess step for insertObserver"


Summary:
We need to separate the step to make the logic more clear
and also to find all the values we want to skip in advance
without the interference of inserted observers

Test Plan:
.

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][refactor] Separate preprocess step for insertObserver"


Summary:
We need to separate the step to make the logic more clear
and also to find all the values we want to skip in advance
without the interference of inserted observers

Test Plan:
.

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][refactor] Separate preprocess step for insertObserver"


Summary:
We need to separate the step to make the logic more clear
and also to find all the values we want to skip in advance
without the interference of inserted observers

Test Plan:
.

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][refactor] Separate preprocess step for insertObserver"


Summary:
We need to separate the step to make the logic more clear
and also to find all the values we want to skip in advance
without the interference of inserted observers

Test Plan:
.

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][refactor] Separate preprocess step for insertObserver"


Summary:
We need to separate the step to make the logic more clear
and also to find all the values we want to skip in advance
without the interference of inserted observers

Test Plan:
.

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][refactor] Separate preprocess step for insertObserver"


Summary:
We need to separate the step to make the logic more clear
and also to find all the values we want to skip in advance
without the interference of inserted observers

Test Plan:
.

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][refactor] Separate preprocess step for insertObserver"


Summary:
We need to separate the step to make the logic more clear
and also to find all the values we want to skip in advance
without the interference of inserted observers

Test Plan:
.

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][refactor] Separate preprocess step for insertObserver"


Summary:
We need to separate the step to make the logic more clear
and also to find all the values we want to skip in advance
without the interference of inserted observers

Test Plan:
.

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][refactor] Separate preprocess step for insertObserver"


Summary:
We need to separate the step to make the logic more clear
and also to find all the values we want to skip in advance
without the interference of inserted observers

Test Plan:
.

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][refactor] Separate preprocess step for insertObserver"


Summary:
We need to separate the step to make the logic more clear
and also to find all the values we want to skip in advance
without the interference of inserted observers

Test Plan:
.

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][refactor] Separate preprocess step for insertObserver"


Summary:
We need to separate the step to make the logic more clear
and also to find all the values we want to skip in advance
without the interference of inserted observers

Test Plan:
.

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][refactor] Separate preprocess step for insertObserver"


Summary:
We need to separate the step to make the logic more clear
and also to find all the values we want to skip in advance
without the interference of inserted observers

Test Plan:
.

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][refactor] Separate preprocess step for insertObserver"


Summary:
We need to separate the step to make the logic more clear
and also to find all the values we want to skip in advance
without the interference of inserted observers

Test Plan:
.

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]
